In [64]:
! pip list

Package         Version   Editable project location
--------------- --------- ------------------------------------------------------------------------------------------------------------
diabetes        0.1.0     /Users/dyada/Desktop/BSE/Computing for Data Science/homeworks/hm5/homework-5-computing-data-science/diabetes
geographiclib   2.0
geopy           2.3.0
joblib          1.2.0
numpy           1.23.4
pandas          1.5.1
pip             22.3.1
python-dateutil 2.8.2
pytz            2022.6
scikit-learn    1.1.3
scipy           1.9.3
setuptools      65.3.0
six             1.16.0
sklearn         0.0.post1
threadpoolctl   3.1.0
wheel           0.37.1


In [65]:
import sys
package_path = '/Users/dyada/Desktop/BSE/Computing for Data Science/homeworks/hm5/homework-5-computing-data-science/diabetes/'
sys.path.append(package_path)

In [66]:
from diabetes.data.load import from_csv
from diabetes.processing.train_test import split
from diabetes.processing.na import drop, fill
from diabetes.processing.dummy import generate, binarize
from diabetes.training.model import train, predict
from diabetes.evaluating.score import roc_auc

In [67]:
X_FEATURES = ['age', 'height', 'weight', 'aids', 'cirrhosis', 'hepatic_failure',
'immunosuppression', 'leukemia', 'lymphoma', 'solid_tumor_with_metastasis']

#### a. Load the data.

In [68]:
df = from_csv(package_path + 'data/sample_diabetes_mellitus_data.csv')

print(df)

      encounter_id  hospital_id   age        bmi  elective_surgery  \
0           214826          118  68.0  22.732803                 0   
1           246060           81  77.0  27.421875                 0   
2           276985          118  25.0  31.952749                 0   
3           262220          118  81.0  22.635548                 1   
4           201746           33  19.0        NaN                 0   
...            ...          ...   ...        ...               ...   
9995        167042          140   NaN  51.952814                 0   
9996        165009           89  34.0  26.110113                 0   
9997        234052          137  30.0  23.774241                 0   
9998        249290           89  67.0  31.330708                 0   
9999        151312           89  55.0  44.343663                 0   

            ethnicity gender  height hospital_admit_source  \
0           Caucasian      M  180.30                 Floor   
1           Caucasian      F  160.0

In [69]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 0 to 9999
Data columns (total 52 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   encounter_id                 10000 non-null  int64  
 1   hospital_id                  10000 non-null  int64  
 2   age                          9595 non-null   float64
 3   bmi                          8262 non-null   float64
 4   elective_surgery             10000 non-null  int64  
 5   ethnicity                    9758 non-null   object 
 6   gender                       9997 non-null   object 
 7   height                       9840 non-null   float64
 8   hospital_admit_source        9082 non-null   object 
 9   icu_admit_source             9979 non-null   object 
 10  icu_id                       10000 non-null  int64  
 11  icu_stay_type                10000 non-null  object 
 12  icu_type                     10000 non-null  object 
 13  pre_icu_los_days 

## c. Remove those rows that contain NaN values in the columns: age, gender, ethnicity.

In [70]:
df = drop(df, columns_l = ['age', 'gender', 'ethnicity'])

print('NaNs dropped in age, gender, ethnicity columns>>>>\n', df)

NaNs dropped in age, gender, ethnicity columns>>>>
       encounter_id  hospital_id   age        bmi  elective_surgery  \
0           214826          118  68.0  22.732803                 0   
1           246060           81  77.0  27.421875                 0   
2           276985          118  25.0  31.952749                 0   
3           262220          118  81.0  22.635548                 1   
4           201746           33  19.0        NaN                 0   
...            ...          ...   ...        ...               ...   
9994        271449           89  78.0  28.700475                 0   
9996        165009           89  34.0  26.110113                 0   
9997        234052          137  30.0  23.774241                 0   
9998        249290           89  67.0  31.330708                 0   
9999        151312           89  55.0  44.343663                 0   

            ethnicity gender  height hospital_admit_source  \
0           Caucasian      M  180.30         

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9368 entries, 0 to 9999
Data columns (total 52 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   encounter_id                 9368 non-null   int64  
 1   hospital_id                  9368 non-null   int64  
 2   age                          9368 non-null   float64
 3   bmi                          7746 non-null   float64
 4   elective_surgery             9368 non-null   int64  
 5   ethnicity                    9368 non-null   object 
 6   gender                       9368 non-null   object 
 7   height                       9241 non-null   float64
 8   hospital_admit_source        8518 non-null   object 
 9   icu_admit_source             9347 non-null   object 
 10  icu_id                       9368 non-null   int64  
 11  icu_stay_type                9368 non-null   object 
 12  icu_type                     9368 non-null   object 
 13  pre_icu_los_days  

## d. Fill NaN with the mean value of the column in the columns: height, weight.

In [72]:
df = fill(df, columns_l = ['height', 'weight'])
print('NaNs filled with means for height, weight columns>>>>\n', df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9368 entries, 0 to 9999
Data columns (total 52 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   encounter_id                 9368 non-null   int64  
 1   hospital_id                  9368 non-null   int64  
 2   age                          9368 non-null   float64
 3   bmi                          7746 non-null   float64
 4   elective_surgery             9368 non-null   int64  
 5   ethnicity                    9368 non-null   object 
 6   gender                       9368 non-null   object 
 7   height                       9368 non-null   float64
 8   hospital_admit_source        8518 non-null   object 
 9   icu_admit_source             9347 non-null   object 
 10  icu_id                       9368 non-null   int64  
 11  icu_stay_type                9368 non-null   object 
 12  icu_type                     9368 non-null   object 
 13  pre_icu_los_days  

## e. Generate dummies for ethnicity column (One hot encoding).

In [73]:
df = generate(df, columns_l = ['ethnicity'])
print('NaNs filled with means for height, weight columns>>>>\n', df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9368 entries, 0 to 9999
Data columns (total 57 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   encounter_id                 9368 non-null   int64  
 1   hospital_id                  9368 non-null   int64  
 2   age                          9368 non-null   float64
 3   bmi                          7746 non-null   float64
 4   elective_surgery             9368 non-null   int64  
 5   gender                       9368 non-null   object 
 6   height                       9368 non-null   float64
 7   hospital_admit_source        8518 non-null   object 
 8   icu_admit_source             9347 non-null   object 
 9   icu_id                       9368 non-null   int64  
 10  icu_stay_type                9368 non-null   object 
 11  icu_type                     9368 non-null   object 
 12  pre_icu_los_days             9368 non-null   float64
 13  readmission_status

### f. Create a binary variable for gender M/F.

In [74]:
df = binarize(df, init_column_name='gender', binarized_column_name='M/F')
print('Binarized gender column to M/F>>>>\n', df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9368 entries, 0 to 9999
Data columns (total 57 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   encounter_id                 9368 non-null   int64  
 1   hospital_id                  9368 non-null   int64  
 2   age                          9368 non-null   float64
 3   bmi                          7746 non-null   float64
 4   elective_surgery             9368 non-null   int64  
 5   height                       9368 non-null   float64
 6   hospital_admit_source        8518 non-null   object 
 7   icu_admit_source             9347 non-null   object 
 8   icu_id                       9368 non-null   int64  
 9   icu_stay_type                9368 non-null   object 
 10  icu_type                     9368 non-null   object 
 11  pre_icu_los_days             9368 non-null   float64
 12  readmission_status           9368 non-null   int64  
 13  weight            

# b. Split the data between train and test. (you can use train_test_split from sklearn or any other way)

In [75]:
X_train, X_test, y_train, y_test = split(data=df, target='diabetes_mellitus')

print(X_train)

      encounter_id  hospital_id   age        bmi  elective_surgery  height  \
4731        194008          118  79.0  40.546531                 0   162.6   
5783        150184           81  85.0        NaN                 0   157.5   
6572        237953           77  46.0        NaN                 1   182.9   
8247        219591          137  88.0  27.302230                 0   175.3   
1516        162123           77  65.0  35.080682                 1   177.8   
...            ...          ...   ...        ...               ...     ...   
2303        191839          118  59.0        NaN                 0   165.1   
8313        177145          137  51.0  34.289412                 0   163.8   
3534        245727          118  59.0  23.526066                 0   162.6   
43          179524           31  52.0  23.039099                 0   165.1   
3549        271596          118  31.0  19.228087                 0   170.2   

     hospital_admit_source           icu_admit_source  icu_id i

In [76]:
print('X_train.shape>>', X_train.shape)
print('X_test.shape>>', X_test.shape)
print('y_train.shape>>', y_train.shape)
print('y_test.shape>>', y_test.shape)

X_train.shape>> (7494, 56)
X_test.shape>> (1874, 56)
y_train.shape>> (7494,)
y_test.shape>> (1874,)


### g. Train a model (for instance LogisticRegression or RandomForestClassifier from sklearn) in the train data. Use as features the columns: ‘age’, ‘height’, ‘weight’, ‘aids’, ‘cirrhosis’, ‘hepatic_failure’, ‘immunosuppression’, ‘leukemia’, ‘lymphoma’, ‘solid_tumor_with_metastasis’. Use as target the column: ‘diabetes_mellitus’

In [77]:
trained_model = train(model='rf', X_train=X_train, y_train=y_train, X_features_l=X_FEATURES)

print('trained_model>>', trained_model)

trained_model>> RandomForestClassifier()


### h. Predict the targets for both the train and test sets and add the prediction as a new column (use predict_proba from the model to get the predicted probabilities) name the new column something like predictions.

In [78]:
y_train, y_test = predict(trained_model, X_train, y_train, X_test, y_test, X_features_l=X_FEATURES)

print('y_train>>>', y_train)
print('y_test>>>', y_test)

y_train>>>       diabetes_mellitus  predictions
4731                  1     0.840000
5783                  0     0.241667
6572                  0     0.644381
8247                  0     0.038571
1516                  1     0.700000
...                 ...          ...
2303                  0     0.077667
8313                  0     0.110000
3534                  0     0.010000
43                    0     0.030000
3549                  0     0.210000

[7494 rows x 2 columns]
y_test>>>       diabetes_mellitus  predictions
6561                  1     0.220000
847                   1     0.570000
6509                  0     0.000000
7459                  0     0.280000
5980                  0     0.373714
...                 ...          ...
2440                  0     0.149167
5578                  0     0.453333
1076                  0     0.315255
3428                  0     0.018333
825                   0     0.246667

[1874 rows x 2 columns]


### i. Compute the train and test roc_auc metric using roc_auc_score from sklearn.

In [79]:
print('AUC ROC score Train: ', roc_auc(y_train['diabetes_mellitus'], y_train['predictions']))
print('AUC ROC score Test: ', roc_auc(y_test['diabetes_mellitus'], y_test['predictions']))

AUC ROC score Train:  0.9981218638135233
AUC ROC score Test:  0.5983142226631561


# clearly Random forest overfits while logistic regression does not if you change model in function trained_model = train(model='rf', X_train=X_train, y_train=y_train, X_features_l=X_FEATURES) to lr

In [80]:
X_train, X_test, y_train, y_test = split(data=df, target='diabetes_mellitus')

print(X_train)

trained_model = train(model='lr', X_train=X_train, y_train=y_train, X_features_l=X_FEATURES)

print('trained_model>>', trained_model)

      encounter_id  hospital_id   age        bmi  elective_surgery  height  \
4731        194008          118  79.0  40.546531                 0   162.6   
5783        150184           81  85.0        NaN                 0   157.5   
6572        237953           77  46.0        NaN                 1   182.9   
8247        219591          137  88.0  27.302230                 0   175.3   
1516        162123           77  65.0  35.080682                 1   177.8   
...            ...          ...   ...        ...               ...     ...   
2303        191839          118  59.0        NaN                 0   165.1   
8313        177145          137  51.0  34.289412                 0   163.8   
3534        245727          118  59.0  23.526066                 0   162.6   
43          179524           31  52.0  23.039099                 0   165.1   
3549        271596          118  31.0  19.228087                 0   170.2   

     hospital_admit_source           icu_admit_source  icu_id i

In [81]:
y_train, y_test = predict(trained_model, X_train, y_train, X_test, y_test, X_features_l=X_FEATURES)

print('y_train>>>', y_train)
print('y_test>>>', y_test)

y_train>>>       diabetes_mellitus  predictions
4731                  1     0.442973
5783                  0     0.390512
6572                  0     0.123382
8247                  0     0.281255
1516                  1     0.299137
...                 ...          ...
2303                  0     0.231314
8313                  0     0.229210
3534                  0     0.152150
43                    0     0.127921
3549                  0     0.064774

[7494 rows x 2 columns]
y_test>>>       diabetes_mellitus  predictions
6561                  1     0.199698
847                   1     0.386706
6509                  0     0.111578
7459                  0     0.165193
5980                  0     0.157344
...                 ...          ...
2440                  0     0.164091
5578                  0     0.206687
1076                  0     0.265802
3428                  0     0.160943
825                   0     0.151273

[1874 rows x 2 columns]


In [82]:
print('AUC ROC score Train: ', roc_auc(y_train['diabetes_mellitus'], y_train['predictions']))
print('AUC ROC score Test: ', roc_auc(y_test['diabetes_mellitus'], y_test['predictions']))

AUC ROC score Train:  0.6698929077986007
AUC ROC score Test:  0.665457543150049


# AUC dropped to around 67% but it is consistent across both train and test sets, also maybe underfitted as well..